src/param_env.md

```diff
@@ -1,69 +1,159 @@
+The type system relies on information in the environment in order for it to function correctly. This information is stored in the [`ParamEnv`][pe] type and it is important to use the correct `ParamEnv` when interacting with the type system.
 
-For example if you have a function
+The information represented by `ParamEnv` is a list of in scope where clauses, and a [`Reveal`][reveal]. A `ParamEnv` typically corresponds to a specific item's environment however it can also be created with arbitrary data that is not derived from a specific item. In most cases `ParamEnv`s are initially created via the [`param_env` query][query] which returns a `ParamEnv` derived from the provided item's where clauses.
 
+If we have a function such as:
 ```rust
-fn foo<T: Copy>(t: T) { ... }
+// `foo` would have a `ParamEnv` of:
+// `[T: Sized, T: Trait, <T as Trait>::Assoc: Clone]`
+fn foo<T: Trait>()
+where
+    <T as Trait>::Assoc: Clone,
+{}
 ```
+If we were interacting with the type system from inside of `foo` we would use this `ParamEnv` everywhere that we interact with the type system. This would allow things such as normalization (TODO: write a chapter about normalization and link it), evaluating generic constants, and proving where clauses/goals, to rely on `T` being sized, implementing `Trait`, etc.
 
-the parameter environment for that function is `[T: Copy]`. This means any
-evaluation within this function will, when accessing the type `T`, know about
-its `Copy` bound via the parameter environment.
+A more concrete example:
+```rust
+// `foo` would have a `ParamEnv` of:
+// `[T: Sized, T: Clone]`
+fn foo<T: Clone>(a: T) -> (T, T) {
+    // when typechecking `foo` we require all the where clauses on `bar`
+    // to hold in order for it to be legal to call. This means we have to
+    // prove `T: Clone`. As we are type checking `foo` we use `foo`'s
+    // environment when trying to check that `T: Clone` holds.
+    //
+    // Trying to prove `T: Clone` with a `ParamEnv` of `[T: Sized, T: Clone]`
+    // will trivially succeed as bound we want to prove is in our environment.
+    requires_clone(a);
+}
+```
 
-You can get the parameter environment for a `def_id` using the
-[`param_env`][query] query. However, this `ParamEnv` can be too generic for
-your use case. Using the `ParamEnv` from the surrounding context can allow you
-to evaluate more things. For example, suppose we had something the following:
+Or alternatively an example that would not compile:
+```rust
+// `foo2` would have a `ParamEnv` of:
+// `[T: Sized]`
+fn foo2<T>(a: T) -> (T, T) {
+    // When typechecking `foo2` we attempt to prove `T: Clone`.
+    // As we are type checking `foo2` we use `foo2`'s environment
+    // when trying to prove `T: Clone`.
+    //
+    // Trying to prove `T: Clone` with a `ParamEnv` of `[T: Sized]` will
+    // fail as there is nothing in the environment telling the trait solver
+    // that `T` implements `Clone`.
```

Comment:
I feel like this should clarified. If we can't find anything in the `ParamEnv` there may still be other candidates like user-written impls. Might sound a bit nit-picky but I think this could fortify the reader's understanding. Of course, there's no such thing as `impl<T> Clone for T { … }` but we might want mention that somehow.

Reply:
no no this is perfect thank you for pointing this out. I remember vaguely thinking of this when writing but it must have slipped my mind thanks <3 I definitely dont want to give the impression that the _only_ way of proving things is via the bounds in `ParamEnv`